### PR TITLE
Add config validation and style tweaks

### PR DESCRIPTION
**Pull request by Kirire**
**Summary of changes:**

1. **Renamed `input_states` to `hidden_states`** in the `torch_forward` method signature and body, to align with HuggingFace Transformers' naming conventions:
   ```python
   def torch_forward(self, hidden_states: torch.Tensor, ...)
   ```
   This change improves consistency with the rest of the codebase and aligns better with the standard used in other HuggingFace models.

2. **Removed redundant attention masking** from the `forward` method:
   ```python
   if attention_mask is not None ...
       hidden_states = (hidden_states * attention_mask[:, :, None]).to(dtype)
   ```
   This operation is already handled later in `torch_forward`, so it's no longer necessary here. Removing it avoids potential duplication and keeps the logic centralized.

3. **Added a configuration validation check in `Mamba2Config.__init__`**:
   ```python
   if hidden_size * expand != num_heads * head_dim:
       raise AttributeError(...)
   ```
   This ensures that the configuration is internally consistent, and helps users catch misconfigurations early by raising a clear error during initialization.

Related to #37554

4. **Improved Loss Handling During Gradient Accumulation**

Modifies the loss computation by replacing the hardcoded use of `CrossEntropyLoss`:

```python
loss_fct = CrossEntropyLoss()
loss = loss_fct(shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1))
```

with a more flexible call to a configurable `loss_function`:

```python
loss = self.loss_function(logits=logits, labels=labels, vocab_size=self.config.vocab_size, **kwargs)
```

This change improves support for gradient accumulation scenarios and allows for easier customization of model-specific loss functions.

Related to #34191

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @vasqu @molbap from #37554

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 1]
```diff
@@ -662,11 +668,6 @@ def forward(
     ):
         if is_fast_path_available and "cuda" in self.in_proj.weight.device.type:
             return self.cuda_kernels_forward(hidden_states, cache_params, cache_position, attention_mask)
-        dtype = hidden_states.dtype
-        if attention_mask is not None and attention_mask.shape[1] > 1 and attention_mask.shape[0] > 1:
-            # tune out hidden states for pad tokens, see https://github.com/state-spaces/mamba/issues/66
-            hidden_states = (hidden_states * attention_mask[:, :, None]).to(dtype)
```
Comment (vasqu) [inline src/transformers/models/mamba2/modeling_mamba2.py:668]: Good catch, this is indeed done beforehand in each forward either way!

**Review by vasqu — APPROVED** [review]
Just a small nit on the type of error to raise. Could you also add a test to check if the error is properly (not) raised? Otherwise LGTM :hugs: 

Also sorry that I came so late to this. In the future, it would help us if you directly ping us in the PR so we don't need redirections / avoid missing your PR @Kirire ;)

**Comment by Kirire** [comment]
Hi! I ran make quality, everything should be good now 👍

**Comment by vasqu** [comment]
Just a last thing: Could you add some test to make sure it works as intended? #33316 is a good reference for how this can be done.

**Comment by vasqu** [comment]
Gentle ping @Kirire

**Review by ArthurZucker — APPROVED** [review]
LGTM thanks 🤗 waiting for the test and we can merge!

**Comment by vasqu** [comment]
@Kirire are you still interested in finishing this PR?

**Comment by Kirire** [comment]
Hi! Sorry for the late reply — I recently became a dad, so things have been a bit busy on my end 👶🙂
But yes, I'm definitely still interested, and I’ll get back to it starting tomorrow!

**Comment by vasqu** [comment]
First of all, congrats to you and your family !!! :partying_face: And no worries, take your time.

**Comment by Kirire** [comment]
Thanks a lot! 😊
I've now added the test for the configuration — let me know if that works for you or if you'd like any changes!

File: tests/models/mamba2/test_modeling_mamba2.py [hunk 2]
```diff
@@ -44,6 +44,33 @@
...
+        self._create_config(hidden_size=4, num_heads=4, expand=2, head_dim=2)
+        self._create_config(hidden_size=2, num_heads=4, expand=4, head_dim=2)
+        with self.parent.assertRaises(ValueError):
+            self._create_config(hidden_size=2, num_heads=2, expand=4, head_dim=4)
+        with self.parent.assertRaises(ValueError):
+            self._create_config(hidden_size=4, num_heads=2, expand=4, head_dim=2)
+
+    def test_mamba2_offset_properties(self):
+        self.test_attn_offsets()
+        self.test_expert_offsets()
```
Comment (vasqu) [inline tests/models/mamba2/test_modeling_mamba2.py:67]:
```suggestion
```
Seems like a copy paste mistake

File: tests/models/mamba2/test_modeling_mamba2.py [hunk 3]
```diff
@@ -44,6 +44,33 @@
...
+        _input_dict["expand"] = expand
+        _input_dict["head_dim"] = head_dim
+        return self.config_class(**_input_dict)
+
+    def test_consistence(self):
+        self._create_config(hidden_size=2, num_heads=2, expand=2, head_dim=2)
+        self._create_config(hidden_size=4, num_heads=4, expand=2, head_dim=2)
+        self._create_config(hidden_size=2, num_heads=4, expand=4, head_dim=2)
+        with self.parent.assertRaises(ValueError):
+            self._create_config(hidden_size=2, num_heads=2, expand=4, head_dim=4)
```
Comment (vasqu) [inline tests/models/mamba2/test_modeling_mamba2.py:61]:
```suggestion
            self._create_config(hidden_size=2, num_heads=2, expand=3, head_dim=4)
```
This shouldn't raise an error, no?

File: tests/models/mamba2/test_modeling_mamba2.py [hunk 4]
```diff
@@ -44,6 +44,33 @@
...
+class Mamba2ConfigTester(ConfigTester):
+    def _create_config(self, hidden_size: int, num_heads: int, expand: int, head_dim: int):
+        _input_dict = self.inputs_dict.copy()
+        _input_dict["hidden_size"] = hidden_size
+        _input_dict["num_heads"] = num_heads
+        _input_dict["expand"] = expand
+        _input_dict["head_dim"] = head_dim
+        return self.config_class(**_input_dict)
+
+    def test_consistence(self):
```
Comment (vasqu) [inline tests/models/mamba2/test_modeling_mamba2.py:56]:
```suggestion
    def test_hidden_size_compatibility(self):
```
Just to have something more descriptive. Should be updated below then in the call :eyes:

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 5]
```diff
@@ -1052,14 +1052,7 @@ def forward(
...
         if labels is not None:
-            # move labels to correct device to enable model parallelism
-            labels = labels.to(logits.device)
-            # Shift so that tokens < n predict n
-            shift_logits = logits[..., :-1, :].contiguous()
-            shift_labels = labels[..., 1:].contiguous()
-            # Flatten the tokens
-            loss_fct = CrossEntropyLoss()
-            loss = loss_fct(shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1))
+            loss = self.loss_function(logits=logits, labels=labels, vocab_size=self.config.vocab_size, **kwargs)
```
Comment (vasqu) [inline src/transformers/models/mamba2/modeling_mamba2.py:1055]: Can you add that change in the PR description? I think it's fine but we should still know what happened in the PR

File: tests/models/mamba2/test_modeling_mamba2.py [hunk 6]
```diff
@@ -44,6 +44,33 @@
...
+            self._create_config(hidden_size=2, num_heads=2, expand=4, head_dim=4)
+        with self.parent.assertRaises(ValueError):
+            self._create_config(hidden_size=4, num_heads=2, expand=4, head_dim=2)
+
+    def test_mamba2_offset_properties(self):
+        self.test_attn_offsets()
+        self.test_expert_offsets()
+
+    def run_common_tests(self):
+        self.test_mamba2_offset_properties()
```
Comment (vasqu) [inline tests/models/mamba2/test_modeling_mamba2.py:70]:
```suggestion
        self.test_hidden_size_compatibility()
```
Or based on a different rename :D

**Review by vasqu — COMMENTED** [review]
Mostly nits / style and some copy/paste mistake

**Comment by Kirire** [comment]
oups my bad ^^'
